### DISCUSSION TITLE
How to process Amazon hosted images.

### DISCUSSION THREAD

**Discussion by qudratullah1061** [libvips/libvips#2409] [Q&A]
Hi,

Thank you for this great library. Really helpful. I came up here with a question. What steps are necessary to do if I want to process images hosted on the amazon s3 bucket with the lipvips command line?

I am executing the command using PHP function exec();
Here is an example.

exec('vips tiffsave "' . $inputPath . '" "' . $outputPath . '" --tile --pyramid ' . $compression . ' ' . $quality . ' --tile-width ' . $tileSize . ' --tile-height ' . $tileSize, $output, $error);

The $inputPath is the path of an image hosted on the same server where I am executing this script. But what if I want to use the amazon s3 bucket image or remote image url? Please help.

Thank you soo much.

**Comment by jcupitt** [comment]
Hi @qudratullah1061,

There's an example here:

https://libvips.github.io/libvips/2019/12/11/What's-new-in-8.9.html#true-streaming

```
aws s3 cp s3://mybucket/input.jpg - | \
  vips thumbnail_source [descriptor=0] .jpg 128 | \
    aws s3 cp - s3://mybucket/output.jpg
```

**Comment by jcupitt** [comment]
... or you can use php-vips, of course, which might be slightly faster.

**Reply by qudratullah1061** [reply]
Hi @jcupitt,

Thanks, What will be my final command for aws hosted image?
Current command. 

vips tiffsave "URL OF IMAGE" "SAVE OUTPUT TO BUCKET" --tile --pyramid --compression none --tile-width 256 --tile-height 256

**Reply by jcupitt** [reply]
Sorry, I hadn't read your post carefully enough. TIFF pyramid creation has to be done on the local filesystem, since internally it's a multi-step process. You'll need to do this in two phases. 

Perhaps:

```
aws s3 cp s3://mybucket/input.jpg - | \
  vips tiffsave stdin x.tif --tile --pyramid --compression none --tile-width 256 --tile-height 256
aws s3 cp x.tif s3://mybucket/output.jpg
```

We've been thinking about ways to fix TIFF creation so it can be done to a pipe, but it no one's implemented it yet, unfortunately.
